src/mvba/tag.rs

```diff
@@ -0,0 +1,42 @@
+use std::fmt::Display;
+
+use super::NodeId;
+
+#[derive(Debug, Clone, Eq, PartialEq, serde::Serialize, serde::Deserialize)]
+pub struct Domain {
+    pub id: String, // this is same as $ID$ in spec
+    pub seq: usize, // this is same as $s$ in spec
+}
+
+impl Domain {
+    pub fn new(id: impl Into<String>, seq: usize) -> Self {
+        Self { id: id.into(), seq }
+    }
+}
+
+impl Display for Domain {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}[{}]", self.id, self.seq)
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, serde::Serialize, serde::Deserialize)]
+pub struct Tag {
```

Comment:
I think Domain here is unnecessary. Just Tag is fair enough, with three parameters:
```rust
pub struct Tag {
   pub id: String,  // this is same as $ID$ in spec
   pub j: NodeId,  // this is same as $j$ in spec
   pub s: usize, // this is same as $s$ in spec
}
```
 ID, proposer_index and s as sequence which right now always is zero.

Reply:
It is not bad idea to add sequence number and proposer but there are some notes:

1. MVBA doesn't need to have proposer; 
2. ABBA doesn't need to have sequence; 
3. In spec VCBC has sequence but it is always set to zero. 
4. We can't use the same consensus for different decisions, because the parties are not the same. 
5. Consensus should be agnostic about domain, it increases its knowledge.

I understand that adding Tag might helps us to have similar thing in above modules, but it might cause some confusion in future. Let me know what do you think.

Reply:
I think you may be missing the point of this PR. We are trying to prevent vote replay attacks.

We will use the same consensus module in many domains. For example, it will be used for membership to track membership changes and in handover to track elder changes.

We want to make sure votes can't be mixed up between these two domains. I.e. prevent membership votes from being used in handover or vice-versa.

To do that we use the `domain.id` to differentiate the instances and prevent votes being replayed.

Next, a domain will likely want to do multiple decision. i.e. membership will want to make a decision on each node that joins or leaves. We need to differentiate votes coming from an old instance of membership so that they can not be replayed. To do that we have sequence numbers.

I grouped these two together in a `Domain` struct, i.e.`Domain {id, seq}`.

Every msg created by any module in consensus needs to include this domain struct to prevent replay attacks.

That's why we need a Domain struct.

Domain let's us differentiate against multiple instances of consensus, Tag lets us differentiate against multiple instances of the sub-modules that run in consensus.

> MVBA does not require a proposer.

I never said it did.

> ABBA does not require a sequence.

It's needed to prevent vote re-use when a domain wants to make multiple decisions in sequence. Otherwise you could take an old ABBA vote and replay it in the next decision

> In the VCBC specification, a sequence is included but it is always set to zero.

That's because they only do one run. They assume the implementer understands the point of the sequence number is to increment it for the next run.

> Different decisions cannot use the same consensus as the parties involved are not the same.

Sorry, that's just wrong. What's stopping the same voters from participating in different domains? We have that in safe network where elders are voting in membership consensus and in handover consensus.

> A consensus should not be biased towards any particular domain, as this increases its versatility.

Nothing about this PR is biasing consensus towards any domain.